### DISCUSSION TITLE
🐛 Custom attributes defined in result_from_instance are not present for default value

### DISCUSSION THREAD

**Discussion by tomas-nz** [codingjoe/django-select2#309] [Q&A]
### Bug Description

Method `result_from_instance` is not called for a default (initial) value. Only method `label_from_instance` is called in this case.

### Steps to Reproduce

If I define a custom widget:

```
class MyWidget(s2forms.ModelSelect2Widget):
    model = MyModel
    queryset = MyModel.objects.all()
    
    def result_from_instance(self, obj, request):
        return {
            'id': obj.pk,
            'text': self.label_from_instance(obj),
            'my_extra': obj.extra_attribute,
        }
```
Then I define a form field with an initial value:
```
my_field = forms.ModelChoiceField(widget=MyWidget, initial=MyModel.objects.get(pk=123))
```
In JavaScript in the browser I look at the selected value with:
```
$('#id_my_field').select2('data')[0]
```
I expect to see attribute `my_extra` but it is not there, only `id` and `text` are. Once I select a different value from the options list, then my extra attribute shows up correctly.

### Expected Behavior

I would expect that `result_from_instance` would be called and my extra attribute would be present when an unchanged default value is queried using `$('#id_my_field').select2('data')[0]`

**Comment by codingjoe** [comment]
Hi Thomas,

Thanks for reaching out. I am not sure, I fully understand your use case yet.
Let me try to explain how the package works. If you specify an initial value, the widget will render an action HTML option-tag. This tag will include a `value` and `innerHtml`.

In fact, this behavior can't easily be adapted. You see, `result_from_instance` gets the request passed from a view. However, Django's form framework, doesn't pass the request around.

That being said, there is always a way. Just not an easy one.

I hope this answer helps a little.

Cheers!
Joe

**Reply by tomas-nz** [reply]
Hi Joe,
Thank you for looking into this. I have two use cases:

1. I need to render different look for the selected item, and for options in the list. I do that by removing class `"django-select2"` and manual initialization in JavaScript, setting up `templateSelection` and `templateResult`. At the moment I override widget's `label_from_instance` returning something like `"selection string" + <separator character> + "option string"`, and split this and use relevant portion in each function. I thought it would be more elegant to have these two parts returned from `result_from_instance` as separate attributes, rather than relying on special separator characters.
2. I need to transport additional attributes for a selected item to be used in my JavaScript code. There are two workarounds I can think of: Pack them into the `label_from_instance` result like above, or deliver them as separate attributes in `result_from_instance`, however for the initial value find a different way of delivery (put them somewhere on the form in a hidden field for example). Again, it would seem more elegant (code which is easier to read and maintain) if data produced by `result_from_instance` would be available for the initial value too.

If you think it is not achievable to change the behavior of `result_from_instance` for initial values, perhaps a note in the documentation explaining how this works would be useful. It could save someone new a bit of time working out these details.
